components/scheduler/executor_executors.go

```diff
@@ -0,0 +1,493 @@
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/resource"
+	"go.temporal.io/server/common/util"
+	"go.temporal.io/server/service/history/hsm"
+	scheduler1 "go.temporal.io/server/service/worker/scheduler"
+	"go.uber.org/fx"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type (
+	ExecutorTaskExecutorOptions struct {
+		fx.In
+
+		Config         *Config
+		MetricsHandler metrics.Handler
+		BaseLogger     log.Logger
+		HistoryClient  resource.HistoryClient
+		FrontendClient workflowservice.WorkflowServiceClient
+	}
+
+	executorTaskExecutor struct {
+		ExecutorTaskExecutorOptions
+	}
+
+	rateLimitedDetails struct {
+		// The requested interval to delay processing by rescheduilng.
+		Delay time.Duration
+	}
+)
+
+const (
+	startWorkflowMinDeadline = 5 * time.Second // Lower bound for the deadline in which buffered actions are dropped.
+	startWorkflowMaxDeadline = 1 * time.Hour   // Upper bound for the deadline in which buffered actions are dropped.
+
+	// Because the catchup window doesn't apply to a manual start, pick a custom
+	// execution deadline before timing out a start.
+	manualStartExecutionDeadline = startWorkflowMaxDeadline
+
+	// Upper bound on how many times starting an individual buffered action should be retried.
+	ExecutorMaxStartAttempts = 10
+
+	errTypeRetryLimitExceeded = "RetryLimitExceeded"
+	errTypeRateLimited        = "RateLimited"
+	errTypeAlreadyStarted     = "serviceerror.WorkflowExecutionAlreadyStarted"
+)
+
+func RegisterExecutorExecutors(registry *hsm.Registry, options ExecutorTaskExecutorOptions) error {
+	e := executorTaskExecutor{
+		ExecutorTaskExecutorOptions: options,
+	}
+	return hsm.RegisterTimerExecutor(registry, e.executeExecuteTask)
+}
+
+func (e executorTaskExecutor) executeExecuteTask(env hsm.Environment, node *hsm.Node, task ExecuteTask) error {
+	schedulerNode := node.Parent
+	scheduler, err := loadScheduler(schedulerNode)
+	if err != nil {
+		return err
+	}
+	logger := newTaggedLogger(e.BaseLogger, scheduler)
+
+	executor, err := e.loadExecutor(node)
+	if err != nil {
+		return err
+	}
+
+	// Make sure we have something to start. If not, we can clear the buffer and
+	// transition to the waiting state.
+	executionInfo := scheduler.Schedule.Action.GetStartWorkflow()
+	if executionInfo == nil || len(executor.BufferedStarts) == 0 {
+		return hsm.MachineTransition(node, func(e Executor) (hsm.TransitionOutput, error) {
+			e.ExecutorInternal.BufferedStarts = nil
+			return TransitionWait.Apply(e, EventWait{
+				Node: node,
+			})
+		})
+	}
+
+	// Drain the Executor's BufferedStarts. drainBuffer will update Executor's
+	// BufferedStarts, as well as Scheduler metadata.
+	e.drainBuffer(logger, env, executor, scheduler, executionInfo)
+
+	// Update SchedulerInfo metadata.
+	err = hsm.MachineTransition(schedulerNode, func(s Scheduler) (hsm.TransitionOutput, error) {
+		s.Info = scheduler.Info
```

Comment:
Why is it important to cache this field?

Reply:
Note that since you're (wrongly) executing all of this under a lock it's safe but this should really not be done that way. You are potentially overwriting this field with a stale value, this can happen if other transitions mutate this field while the lock is released.